src/helpers/defineMessagesAnalyzer.ts

```diff
@@ -0,0 +1,115 @@
+import { Rule } from "eslint";
+import * as ESTree from 'estree';
+import { Dictionary, CallExpressionNode } from "./types";
+
+type TrackedMessage = {
+  isReported: boolean,
+  node: ESTree.Property,
+  context: Rule.RuleContext,
+}
+
+type MessageId = string | number;
+
+export default class DefineMessagesDuplicationAnalyzer {
+  private idTracker: Dictionary<MessageId, TrackedMessage> = {};
+  private context?: Rule.RuleContext;
+
+  private setContext = (context: Rule.RuleContext): void => {
+    this.context = context;
+  }
+
+  private reportDuplication = (
+    messageId: MessageId, context: Rule.RuleContext, messageNode: ESTree.Property
+  ): void  => {
+    context.report({
+      message: `message with id '${messageId}' is duplicated`,
+      node: messageNode,
+    });
+  }
+
+  private getMessageId = (messageNode: ESTree.Property): number | string | null => {
+    const messageNodeProperies: ESTree.Property[] = messageNode.value.type === 'ObjectExpression' ?
+      this.removeSpreadElements(messageNode.value.properties) : [];
+
+    const messageIdNode = messageNodeProperies.find((messagesProperty: ESTree.Property) => messagesProperty.key.type === 'Identifier' && messagesProperty.key.name === 'id');
+
+    if ((messageIdNode == null) || messageIdNode.value.type !== 'Literal') {
+      return null;
+    }
+
+    const messageId = messageIdNode.value.value;
+    if (typeof messageId === 'number' || typeof messageId === 'string') {
+      return messageId;
+    }
+
+    return null;
+  }
+
+  private checkMessageDuplication = (messageNode: ESTree.Property): void => {
+    const messageId: string | number | null = this.getMessageId(messageNode);
+    if (messageId == null || (this.context == null)) {
+      return;
+    }
+
+    const firstTrakedMessage = this.idTracker[messageId];
+    if (firstTrakedMessage == null) {
+      this.idTracker[messageId] = {
+        node: messageNode,
+        context: this.context,
+        isReported: false,
+      };
+
+      return;
+    }
+
+    this.reportDuplication(messageId, this.context, messageNode);
+
+    if (!firstTrakedMessage.isReported) {
+      firstTrakedMessage.isReported = true;
+      this.reportDuplication(messageId, firstTrakedMessage.context, firstTrakedMessage.node);
+    }
+  }
+
+  public proceedDefineMessagesFunctionCall = (node: CallExpressionNode, context: Rule.RuleContext): void => {
+    this.setContext(context);
+    const isDefineMessagesFunctionCall = this.getIsDefineMessagesFunctionNode(node);
+    if (!isDefineMessagesFunctionCall) {
+      return;
+    }
+
+    const messageNodeList = this.getMessageNodeList(node);
+    messageNodeList.forEach(this.checkMessageDuplication);
```

Comment:
Now the only place where you read `this.context` is `checkMessageDuplication`. So, you can also pass it to that function and then remove this.context completely.
```suggestion
    messageNodeList.forEach((node) => {this.checkMessageDuplication(node, context)});
```